air/src/execution_step/value_types/canon_stream_map.rs

```diff
@@ -0,0 +1,280 @@
+use crate::ExecutionError;
+use crate::JValue;
+use crate::StreamMapKeyError::UnsupportedKVPairObjectOrMapKeyType;
+use crate::UncatchableError;
+
+use air_interpreter_cid::CID;
+use air_interpreter_data::CanonResultCidAggregate;
+use polyplets::SecurityTetraplet;
+
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::ops::Deref;
+use std::rc::Rc;
+
+/// Canon stream map is a read-only struct that mimics conventional map.
+/// The contents of a map are fixed at a certain node.
+#[derive(Debug, Clone)]
+pub struct CanonStreamMap<'key> {
+    // values vec contains all KVpair objects {"key": key, "value": value}.
+    // There might be multiple KVPars with the same key.
+    values: Vec<ValueAggregate>,
+    // Index access leverages the map that does key to CanonStream mapping.
+    map: HashMap<StreamMapKey<'key>, CanonStream>,
+    // ap arg processing leverages this tetraplet
+    tetraplet: Rc<SecurityTetraplet>,
+}
+
+impl<'key> CanonStreamMap<'key> {
+    // The argument's tetraplet is used to produce canon streams for keys so
+    // that the produced canon streams share tetraplets with the original canon stream
+    // rendered by canon instruction.
+    pub(crate) fn from_canon_stream(canon_stream: CanonStream) -> ExecutionResult<CanonStreamMap<'key>> {
+        let mut map: HashMap<StreamMapKey<'_>, CanonStream> = HashMap::new();
+        let tetraplet = canon_stream.tetraplet.clone();
+        for kvpair_obj in canon_stream.iter() {
+            let key = StreamMapKey::from_kvpair(kvpair_obj.clone())
+                .ok_or(UncatchableError::StreamMapKeyError(UnsupportedKVPairObjectOrMapKeyType))?;
+
+            let value = get_value_from_obj(kvpair_obj)?;
+            let entry_canon_stream = map
+                .entry(key)
+                .or_insert(CanonStream::new(vec![], canon_stream.tetraplet().clone()));
+            entry_canon_stream.push(&value);
+        }
+        let values = canon_stream.get_values();
+        Ok(Self { values, map, tetraplet })
+    }
+
+    // This returns a number of distinct keys in the canon map.
+    pub(crate) fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    pub(crate) fn is_empty(&self) -> bool {
+        self.map.is_empty()
+    }
+
+    pub(crate) fn as_jvalue(&self) -> JValue {
+        // TODO: this clone will be removed after boxed values
+        let jvalue_array = self
+            .values
+            .iter()
+            .map(|r| r.get_result().deref().clone())
+            .collect::<Vec<_>>();
+        JValue::Array(jvalue_array)
+    }
+
+    pub(crate) fn iter(&self) -> impl ExactSizeIterator<Item = &ValueAggregate> {
+        self.values.iter()
+    }
+
+    pub(crate) fn tetraplet(&self) -> &Rc<SecurityTetraplet> {
+        &self.tetraplet
+    }
+
+    pub(crate) fn index<'self_l>(&'self_l self, stream_map_key: &StreamMapKey<'key>) -> Option<&'self_l CanonStream> {
+        self.map.get(stream_map_key)
+    }
+
+    #[allow(dead_code)]
```

Comment:
As I understand, this function is intended only for tests, let's decouple it into the test mod (test code located in a file where a structure defined has a full access to its private fields) to not to use it accidentally.